megatron/model/transformer.py

```diff
@@ -429,59 +435,101 @@ def flash_attention(self, query_layer, key_layer, value_layer):
         )
 
         if self.pos_emb != "alibi":
-            # [s, b, np, hn] -> [b, s, np, hn] -> [b * s, 1, np, hn]
-            query_layer = query_layer.transpose(0, 1).reshape(
-                output_size[0] * output_size[2], 1, output_size[1], -1
-            )
+
+            # [sk, b, np, hn] -> [b, sk, np, hn] -> [b * sk, 1, np, hn]
             key_layer = key_layer.transpose(0, 1).reshape(
                 output_size[0] * output_size[3], 1, output_size[1], -1
             )
             value_layer = value_layer.transpose(0, 1).reshape(
                 output_size[0] * output_size[3], 1, output_size[1], -1
             )
-            # Combined q/k/v into [b * s, 3, np, hn].
-            qkv = torch.concat([query_layer, key_layer, value_layer], dim=1)
 
             batch_size = output_size[0]
-            seqlen = output_size[2]
-            max_s = seqlen
+            max_seqlen_q = output_size[2]
+            max_seqlen_k = output_size[3]
 
-            cu_seqlens = torch.arange(
+            cu_seqlens_q = torch.arange(
                 0,
-                (batch_size + 1) * seqlen,
-                step=seqlen,
+                (batch_size + 1) * max_seqlen_q,
+                step=max_seqlen_q,
                 dtype=torch.int32,
-                device=qkv.device,
+                device=query_layer.device,
             )
 
-            output = self.flash_attention_function(
-                qkv,
-                cu_seqlens,
-                max_s,
-                self.dropout_p if self.training else 0.0,
-                softmax_scale=None,
-                causal=True,
+            cu_seqlens_k = torch.arange(
+                0,
+                (batch_size + 1) * max_seqlen_k,
+                step=max_seqlen_k,
+                dtype=torch.int32,
+                device=key_layer.device,
             )
+
+            if self.training:
+
+                # [sq, b, np, hn] -> [b * sq, np, hn]
+                query_layer = query_layer.transpose(0, 1).reshape(
+                    output_size[0] * output_size[2], output_size[1], -1
+                )
+
+                # Combined k/v into [b * sk, 2, np, hn].
+                kv = torch.concat([key_layer, value_layer], dim=1)
+
+                output = self.flash_attn_unpadded_kvpacked_func(
```

Comment:
this needs to be `self.flash_attention_function`

Reply:
yeah. Did you attempted doing inference? If not, I'd do it and push if any more changes.

Reply:
Will run the inference and push any changes as needed.